### ISSUE TITLE
Potential race condition in ChangeTrackingSessionStateItemCollection.cs:GetSessionNormalizedKeyToUse

### ISSUE THREAD

**Issue by vgheri** [Azure/aspnet-redis-providers#48]
Hello,

I work on a heavy traffic website and I canary deployed the redis session state provider on one webserver.
As soon as I deployed it I noticed errors popping up in my log: 

> "System.ArgumentException","Message":"An item with the same key has already been added" 

and the stack trace says:
`"StackTraceString":" at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)\r\n at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)\r\n at Microsoft.Web.Redis.ChangeTrackingSessionStateItemCollection.GetSessionNormalizedKeyToUse(String name) in d:\\TeamCityBuildAgent\\work\\f55792526e6d9089\\src\\Shared\\ChangeTrackingSessionStateItemCollection.cs:line 33\r\n at Microsoft.Web.Redis.ChangeTrackingSessionStateItemCollection.get_Item(String name) in d:\\TeamCityBuildAgent\\work\\f55792526e6d9089\\src\\Shared\\ChangeTrackingSessionStateItemCollection.cs:line 142\r\n`

The result was that I got requests hanging indefinetely in the app pool worker process, causing the app pool to use lots of CPU and had to rollback changes.

This is happening on the home page, when the user requests the page and the JS in the page performs an Ajax requests to load some data.
Both requests pass through some base controller logic which uses the session (same key) and which from time to time causes this race condition when a thread tries to find an entry and another write.

I noticed that the dictionary used in the ChangeTrackingSessionStateItemCollection.cs is not a thread safe data structure, so maybe that's the reason?
Please have a look at [this ](http://stackoverflow.com/questions/33153485/is-it-possible-for-a-dictionary-in-net-to-cause-dead-lock-when-reading-and-writ) Stack Overflow question for more info on why the race condition.

This is a screenshot of an analyzed set of data from PerfView which shows clearly the same case as  the one describe in the SO question above:
![image](https://cloud.githubusercontent.com/assets/1318563/19604715/bf3de33e-97b6-11e6-8cdd-eeca2a9a0443.png)

What do you think?

**Comment by vgheri** [comment]
Hello, is this even still officially supported?

**Comment by JonCole** [comment]
Yes, this is still supported.  We haven't yet had a chance to investigate but hope to investigate soon.  If you wanted to send a pull request with the fix, then that might speed things up.

**Comment by helioraptor** [comment]
sent request #52

**Comment by SiddharthChatrolaMs** [comment]
Can you please explain why this would be a race condition? Every request has their own copy of "ChangeTrackingSessionStateItemCollection" object. so why would there be a race condition? simple test app would be very helpful here.

**Comment by iPipelineMK** [comment]
I have the exact same issue parallel tasks updating the same session item causing the "An item with the same key has already been added" to be thrown in my application.

**Comment by JonCole** [comment]
We will take a look at fixing this.  If someone sent a PR, it would likely speed up the release timeline...

**Comment by iPipelineMK** [comment]
Here is a pull request: https://github.com/Azure/aspnet-redis-providers/pull/95 
Thanks!

**Comment by JonCole** [comment]
So, after further consideration, we decided to double check with the ASP.NET team about this scenario and they said this is a bad idea. 

They basically said that ASP.NET has tight restrictions in that you shouldn’t consume the “god objects” (HttpContext and friends) from a background thread, even if you’re manually ensuring serialized access to these objects, as the god objects reference thread-local storage and other non-shareable context. In practice people do occasionally manually serialize access and call these objects from background threads, and this does _sometimes_ work depending on the scenario.

Basically, this is not a supported scenario and even if we make this change, there is a pretty good chance that you will run into other problems.

**Comment by Xerillio** [comment]
I just ran into this issue as well and thought I'd chime in. @JonCole you say ASP.NET has restrictions on consuming HttpContext from background threads. Does that mean both for read and write access?
We are doing some background processing which only needs read access and with that in mind it smells  to me a bit like bad design that this cannot be done in a thread safe way. Not sure if this on the ASP.NET side or the Redis provider, but would it still not be considered a "bug" that you cannot get thread safe read access?
